### PR TITLE
Fix DTS integration test to read Bicep content from workspace filesystem

### LINKED ISSUE #1724: Integration test failure: azure-prepare – Durable Task Scheduler Bicep not via create tool [Assertion mismatch]
## Diagnosis

The `generates Durable Task Scheduler infrastructure and workflow code for a workflow app` test fails because the agent creates Bicep files using **bash heredoc commands** (`cat > file.bicep << 'EOF'`) rather than the MCP `create` tool. The test collects Bicep content only from `create` MCP tool calls, so it sees `bicepContent.length === 0`.

### Failure

| Field | Value |
|-------|-------|
| **Test** | `generates Durable Task Scheduler infrastructure and workflow code for a workflow app` |
| **Test Class** | `azure-prepare_ - Integration Tests › durable-task-scheduler` |
| **File** | `tests/azure-prepare/integration.test.ts:1008` |
| **Assertion** | `expect(bicepContent.length).toBeGreaterThan(0)` |
| **Received** | `0` — no `.bicep` files found via `create` tool calls |

### Root Cause

The test uses:
````typescript
const createCalls = getToolCalls(agentMetadata, "create");
const bicepContents = createCalls
  .filter(event => args?.path?.endsWith(".bicep"))
  .map(event => args?.file_text ?? "");
const bicepContent = bicepContents.join("\n");
expect(bicepContent.length).toBeGreaterThan(0); // line 1008
```

Analysis of the agent metadata shows the agent created files using bash (29 bash tool calls vs. 2 create tool calls), with only `deployment-plan.md` written via the `create` tool. Bicep files (`infra/main.bicep`, `infra/app/durable-task-scheduler.bicep`) were written via bash heredocs.

Subsequent assertions also fail as a result:
- `expect(/Microsoft\.DurableTask\/schedulers/i.test(bicepContent)).toBe(true)` → false
- `expect(/Microsoft\.DurableTask\/schedulers\/taskHubs/i.test(bicepContent)).toBe(true)` → false
- `expect(/0ad04412-c4d5-4796-b79c-f76d14c8d402/i.test(bicepContent)).toBe(true)` → false

### Prompt Context

```
Prepare the Azure deployment infrastructure for a new workflow app that will orchestrate
a multi-step order processing pipeline. Generate the Bicep templates, RBAC assignments,
and azure.yaml. Use the eastus2 region and my current subscription.
````

### azure-prepare Skill Invocation

azure-prepare was invoked ✅. The failure is in how the agent creates output files (bash vs. create tool).

### Skill Report Context

From the SKILL-REPORT (Test 33):
- Agent reported **95% confidence** and "PASSED" — it generated all required files including DTS scheduler Bicep, TypeScript orchestrator code, RBAC assignments, and azure.yaml
- Files listed: `infra/app/durable-task-scheduler.bicep`, `src/functions/orderProcessingOrchestrator.ts`, `host.json`
- The skill completed the task correctly; the test could not verify Bicep content because files were written via bash

### Environment

| Field | Value |
|-------|-------|
| **Run** | [23978564120](https://github.com/microsoft/GitHub-Copilot-for-Azure/actions/runs/23978564120) |
| **Branch** | main (`a3bdc6e`) |
| **Date** | 2026-04-04 |
| **Model** | claude-sonnet-4.5 |

### Expected Fix

Options to investigate:
1. **Preferred:** Instruct the agent to use the `create` MCP tool (not bash) when writing Bicep and other infrastructure files, so that test assertions on tool call arguments work correctly
2. **Alternative:** Update test to inspect the workspace filesystem directly instead of relying solely on `create` tool calls

> Generated by [Analyze Test Run](https://github.com/microsoft/GitHub-Copilot-for-Azure/actions/runs/23980905110/agentic_workflow) · ● 3.7M · [◷](https://github.com/search?q=repo%3Amicrosoft%2FGitHub-Copilot-for-Azure+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2FGitHub-Copilot-for-Azure%2Fanalyze-test-run%22&type=issues)

### PR DESCRIPTION
**Pull request by greenie-msft**
## Summary

The `generates Durable Task Scheduler infrastructure and workflow code` test failed because it collected Bicep content only from `create` MCP tool calls, but the agent wrote `.bicep` files via bash heredocs instead (`bicepContent.length === 0`).

## Fix

Changed the test to read Bicep files directly from the workspace filesystem using `listFilesRecursive` + `fs.readFileSync`, which works regardless of the file creation mechanism (create tool, bash, etc.).

This is the same pattern already used by other utilities in `tests/azure-prepare/utils.ts` (e.g., `getDockerContext`, `hasServicesSection`).

Fixes #1724

### CODE REVIEW

File: tests/azure-prepare/integration.test.ts [hunk 1]
```diff
@@ -991,20 +992,12 @@ describeIntegration(`${SKILL_NAME}_ - Integration Tests`, () => {
...
-          })
-          .map(event => {
-            const args = (event.data as Record<string, unknown>).arguments as { file_text?: string };
-            return args?.file_text ?? "";
-          });
-        const bicepContent = bicepContents.join("\n");
+        // Read Bicep file contents from the workspace filesystem.
+        // The agent may create files via the `create` tool or bash heredocs,
+        // so reading from disk is more reliable than inspecting tool calls.
+        const allFiles = listFilesRecursive(workspacePath!);
```
Comment (jongio) [inline tests/azure-prepare/integration.test.ts]: `doesWorkspaceFileIncludePattern` in `evaluate.ts` already does this - reads workspace files, skips `node_modules` and directories, wraps reads in try/catch. The deploy test uses it for the exact same DTS bicep assertions (`azure-deploy/integration.test.ts:358-361`):

```typescript
const bicepPattern = /\.bicep$/;
expect(doesWorkspaceFileIncludePattern(workspacePath!, /Microsoft\.DurableTask\/schedulers/i, bicepPattern)).toBe(true);
expect(doesWorkspaceFileIncludePattern(workspacePath!, /Microsoft\.DurableTask\/schedulers\/taskHubs/i, bicepPattern)).toBe(true);
expect(doesWorkspaceFileIncludePattern(workspacePath!, /0ad04412-c4d5-4796-b79c-f76d14c8d402/i, bicepPattern)).toBe(true);
expect(doesWorkspaceFileIncludePattern(workspacePath!, /ipAllowlist/i, bicepPattern)).toBe(true);
expect(doesWorkspaceFileIncludePattern(workspacePath!, /DURABLE_TASK_SCHEDULER_CONNECTION_STRING/i)).toBe(true);
```

This would fix both edge-case concerns from the other review (isFile filtering, node_modules exclusion), drop the `fs` import, and match the established pattern. The `listFilesRecursive` import wouldn't be needed either.

**Review by jongio — COMMENTED** [review]
The approach is right - reading from disk instead of inspecting tool calls makes the test agnostic to how the agent creates files. One suggestion below.

**Review by jongio — CHANGES_REQUESTED** [review]
The fix direction is right - reading from disk instead of tool calls makes this test agnostic to how the agent creates files.

The main issue is that `listFilesRecursive` + `readFileSync` skips edge-case handling that already exists in `doesWorkspaceFileIncludePattern` (isFile check, node_modules exclusion, try/catch). The deploy test uses this exact helper for the same DTS Bicep assertions.

The PR already uses it for the `DURABLE_TASK_SCHEDULER_CONNECTION_STRING` check - extending it to all assertions would drop the `fs` and `listFilesRecursive` imports and match the established pattern. The `ipAllowlist` assertion below the diff also depends on `bicepContent` and needs the same treatment.

**Review by jongio — APPROVED** [review]
Looks good - all assertions now use `doesWorkspaceFileIncludePattern` consistently, matching the deploy test pattern. Unused imports removed, `ipAllowlist` handled the same way. Clean simplification.

**Review by JasonYeMSFT — APPROVED** [review]

File: tests/azure-prepare/integration.test.ts [hunk 2]
```diff
@@ -991,20 +992,12 @@ describeIntegration(`${SKILL_NAME}_ - Integration Tests`, () => {
...
+        const bicepFiles = allFiles.filter(f => f.endsWith(".bicep"));
+        const bicepContent = bicepFiles.map(f => fs.readFileSync(f, "utf-8")).join("\n");
```
[the remark this answers is not in the document]
Comment (JasonYeMSFT) [reply tests/azure-prepare/integration.test.ts]: We can safely assume no directory's name ends with ".bicep" and no other kinds of non-regular file is present.

File: tests/azure-prepare/integration.test.ts [hunk 3]
```diff
@@ -1017,11 +1010,9 @@ describeIntegration(`${SKILL_NAME}_ - Integration Tests`, () => {
...
 
         // Must include the scheduler connection string app setting
-        const allFileContents = createCalls
-          .map(event => {
-            const args = (event.data as Record<string, unknown>).arguments as { file_text?: string };
-            return args?.file_text ?? "";
-          })
+        const allFileContents = allFiles
+          .filter(f => !fs.statSync(f).isDirectory())
+          .map(f => fs.readFileSync(f, "utf-8"))
```
[the remark this answers is not in the document]
Comment (JasonYeMSFT) [reply tests/azure-prepare/integration.test.ts]: The helper function doesWorkspaceFileIncludePattern has already implemented the steps to skip searching in "node_modules" directories. And this function is not expected to throw any error.
